File: src/browserstack.ts
/**
 * BrowserStack App Automate integration for mobile-mcp
 *
 * This module provides a Robot implementation that connects to BrowserStack's
 * Appium hub to control remote real devices.
 *
 * Required environment variables:
 *   BROWSERSTACK_USERNAME - Your BrowserStack username
 *   BROWSERSTACK_ACCESS_KEY - Your BrowserStack access key
 *
 * Optional environment variables:
 *   BROWSERSTACK_APP_URL - URL of uploaded app (bs://xxx or app_url)
 */

import { ActionableError, Button, InstalledApp, Orientation, Robot, ScreenElement, ScreenSize, SwipeDirection } from "./robot";

const BROWSERSTACK_HUB = "https://hub-cloud.browserstack.com/wd/hub";
const BROWSERSTACK_API = "https://api-cloud.browserstack.com";

export interface BrowserStackCredentials {
	username: string;
	accessKey: string;
}

export interface BrowserStackDevice {
	id: string;
	device: string;
	os: string;
	os_version: string;
	realMobile: boolean;
}

export interface BrowserStackSession {
	sessionId: string;
	device: string;
	os: string;
	os_version: string;
}

/**
 * Get BrowserStack credentials from environment variables
 */
export function getBrowserStackCredentials(): BrowserStackCredentials | null {
	const username = process.env.BROWSERSTACK_USERNAME;
	const accessKey = process.env.BROWSERSTACK_ACCESS_KEY;

	if (!username || !accessKey) {
		return null;
	}

	return { username, accessKey };
}

/**
 * Check if BrowserStack is configured
 */
export function isBrowserStackConfigured(): boolean {
	return getBrowserStackCredentials() !== null;
}

/**
 * BrowserStack Device Manager - lists available devices and manages sessions
 */
export class BrowserStackManager {
	private credentials: BrowserStackCredentials;

	constructor(credentials: BrowserStackCredentials) {
		this.credentials = credentials;
	}

	private getAuthHeader(): string {
		const auth = Buffer.from(`${this.credentials.username}:${this.credentials.accessKey}`).toString("base64");
		return `Basic ${auth}`;
	}

	/**
	 * Get list of available devices from BrowserStack
	 */
	public async getAvailableDevices(): Promise<BrowserStackDevice[]> {
		const response = await fetch(`${BROWSERSTACK_API}/app-automate/devices.json`, {
			headers: {
				"Authorization": this.getAuthHeader(),
			},
		});

		if (!response.ok) {
			throw new ActionableError(`Failed to fetch BrowserStack devices: ${response.status}`);
		}

		const devices = await response.json() as Array<{
			device: string;
			os: string;
			os_version: string;
			realMobile: boolean;
		}>;

		return devices.map(d => ({
			id: `browserstack:${d.os}:${d.device}:${d.os_version}`,
			device: d.device,
			os: d.os,
			os_version: d.os_version,
			realMobile: d.realMobile,
		}));
	}

	/**
	 * Create a new Appium session on BrowserStack
	 */
	public async createSession(device: string, os: string, osVersion: string, appUrl?: string): Promise<string> {
		const capabilities: Record<string, any> = {
			"bstack:options": {
				userName: this.credentials.username,
				accessKey: this.credentials.accessKey,
				deviceName: device,
				osVersion: osVersion,
				realMobile: "true",
				local: "false",
				debug: "true",
				networkLogs: "true",
			},
			"platformName": os === "ios" ? "iOS" : "Android",
		};

		// Add app URL if provided
		const app = appUrl || process.env.BROWSERSTACK_APP_URL;
		if (app) {
			capabilities["bstack:options"].appUrl = app;
		}

		const response = await fetch(`${BROWSERSTACK_HUB}/session`, {
			method: "POST",
			headers: {
				"Content-Type": "application/json",
				"Authorization": this.getAuthHeader(),
			},
			body: JSON.stringify({
				capabilities: {
					alwaysMatch: capabilities,
				},
			}),
		});

		if (!response.ok) {
			const errorText = await response.text();
			throw new ActionableError(`Failed to create BrowserStack session: ${response.status} ${errorText}`);
		}

		const result = await response.json() as { value: { sessionId: string } };
		return result.value.sessionId;
	}

	/**
	 * Delete a BrowserStack session
	 */
	public async deleteSession(sessionId: string): Promise<void> {
		await fetch(`${BROWSERSTACK_HUB}/session/${sessionId}`, {
			method: "DELETE",
			headers: {
				"Authorization": this.getAuthHeader(),
			},
		});
	}

	/**
	 * Upload an app to BrowserStack
	 */
	public async uploadApp(filePath: string): Promise<string> {
		const fs = await import("node:fs");
		const path = await import("node:path");

		const fileBuffer = fs.readFileSync(filePath);
		const fileName = path.basename(filePath);

		// Create multipart form data manually
		const boundary = `----WebKitFormBoundary${Date.now()}`;
		const body = Buffer.concat([
			Buffer.from(`--${boundary}\r\n`),
			Buffer.from(`Content-Disposition: form-data; name="file"; filename="${fileName}"\r\n`),
			Buffer.from(`Content-Type: application/octet-stream\r\n\r\n`),
			fileBuffer,
			Buffer.from(`\r\n--${boundary}--\r\n`),
		]);

		const response = await fetch(`${BROWSERSTACK_API}/app-automate/upload`, {
			method: "POST",
			headers: {
				"Authorization": this.getAuthHeader(),
				"Content-Type": `multipart/form-data; boundary=${boundary}`,
			},
			body: body,
		});

		if (!response.ok) {
			const errorText = await response.text();
			throw new ActionableError(`Failed to upload app to BrowserStack: ${response.status} ${errorText}`);
		}

		const result = await response.json() as { app_url: string };
		return result.app_url;
	}
}

/**
 * BrowserStack Robot - controls a remote device via BrowserStack's Appium hub
 */
export class BrowserStackRobot implements Robot {
	private credentials: BrowserStackCredentials;
	private sessionId: string;
	private hubUrl: string;

	constructor(credentials: BrowserStackCredentials, sessionId: string) {
		this.credentials = credentials;
		this.sessionId = sessionId;
		this.hubUrl = BROWSERSTACK_HUB;
	}

	private getAuthHeader(): string {
		const auth = Buffer.from(`${this.credentials.username}:${this.credentials.accessKey}`).toString("base64");
		return `Basic ${auth}`;
	}

	private async request(method: string, path: string, body?: any): Promise<any> {
		const url = `${this.hubUrl}/session/${this.sessionId}${path}`;
		const response = await fetch(url, {
			method,
			headers: {
				"Content-Type": "application/json",
				"Authorization": this.getAuthHeader(),
			},
			body: body ? JSON.stringify(body) : undefined,
		});

		if (!response.ok) {
			const errorText = await response.text();
			throw new ActionableError(`BrowserStack request failed: ${response.status} ${errorText}`);
		}

		const text = await response.text();
		if (!text) {return null;}

		return JSON.parse(text);
	}

	public async getScreenSize(): Promise<ScreenSize> {
		const result = await this.request("GET", "/window/rect");
		return {
			width: result.value.width,
			height: result.value.height,
			scale: 1, // BrowserStack returns actual pixels
		};
	}

	public async swipe(direction: SwipeDirection): Promise<void> {
		const screenSize = await this.getScreenSize();
		const centerX = Math.floor(screenSize.width / 2);
		const centerY = Math.floor(screenSize.height / 2);
		const distance = Math.floor(screenSize.height * 0.4);

		let startX = centerX, startY = centerY, endX = centerX, endY = centerY;

		switch (direction) {
			case "up":
				startY = centerY + distance / 2;
				endY = centerY - distance / 2;
				break;
			case "down":
				startY = centerY - distance / 2;
				endY = centerY + distance / 2;
				break;
			case "left":
				startX = centerX + distance / 2;
				endX = centerX - distance / 2;
				break;
			case "right":
				startX = centerX - distance / 2;
				endX = centerX + distance / 2;
				break;
		}

		await this.performSwipe(startX, startY, endX, endY);
	}

	public async swipeFromCoordinate(x: number, y: number, direction: SwipeDirection, distance: number = 400): Promise<void> {
		let endX = x, endY = y;

		switch (direction) {
			case "up":
				endY = y - distance;
				break;
			case "down":
				endY = y + distance;
				break;
			case "left":
				endX = x - distance;
				break;
			case "right":
				endX = x + distance;
				break;
		}

		await this.performSwipe(x, y, endX, endY);
	}

	private async performSwipe(startX: number, startY: number, endX: number, endY: number): Promise<void> {
		await this.request("POST", "/actions", {
			actions: [
				{
					type: "pointer",
					id: "finger1",
					parameters: { pointerType: "touch" },
					actions: [
						{ type: "pointerMove", duration: 0, x: startX, y: startY },
						{ type: "pointerDown", button: 0 },
						{ type: "pointerMove", duration: 1000, x: endX, y: endY },
						{ type: "pointerUp", button: 0 },
					],
				},
			],
		});

		// Clear actions
		await this.request("DELETE", "/actions");
	}

	public async getScreenshot(): Promise<Buffer> {
		const result = await this.request("GET", "/screenshot");
		return Buffer.from(result.value, "base64");
	}

	public async listApps(): Promise<InstalledApp[]> {
		// BrowserStack doesn't provide an API to list installed apps
		// Return empty array - user should know which apps they've uploaded
		return [];
	}

	public async launchApp(packageName: string, _locale?: string): Promise<void> {
		await this.request("POST", "/appium/device/activate_app", {
			bundleId: packageName, // iOS
			appPackage: packageName, // Android - Appium handles both
		});
	}

	public async terminateApp(packageName: string): Promise<void> {
		await this.request("POST", "/appium/device/terminate_app", {
			bundleId: packageName,
			appPackage: packageName,
		});
	}

	public async installApp(path: string): Promise<void> {
		// For BrowserStack, apps must be uploaded first via the API
		// The path should be a bs:// URL
		if (path.startsWith("bs://")) {
			await this.request("POST", "/appium/device/install_app", {
				appPath: path,
			});
		} else {
			throw new ActionableError(
				"BrowserStack requires apps to be uploaded first. Use the BrowserStack API to upload your app and get a bs:// URL."
			);
		}
	}

	public async uninstallApp(bundleId: string): Promise<void> {
		await this.request("POST", "/appium/device/remove_app", {
			bundleId: bundleId,
			appPackage: bundleId,
		});
	}

	public async openUrl(url: string): Promise<void> {
		await this.request("POST", "/url", { url });
	}

	public async sendKeys(text: string): Promise<void> {
		// Find the active element and send keys to it
		const activeElement = await this.request("POST", "/element/active");
		if (activeElement?.value?.ELEMENT) {
			await this.request("POST", `/element/${activeElement.value.ELEMENT}/value`, {
				value: text.split(""),
			});
		} else {
			// Fallback: use keyboard actions
			await this.request("POST", "/actions", {
				actions: [{
					type: "key",
					id: "keyboard",
					actions: text.split("").flatMap(char => [
						{ type: "keyDown", value: char },
						{ type: "keyUp", value: char },
					]),
				}],
			});
		}
	}

	public async pressButton(button: Button): Promise<void> {
		const buttonMap: Record<string, string> = {
			"HOME": "home",
			"BACK": "back",
			"VOLUME_UP": "volumeup",
			"VOLUME_DOWN": "volumedown",
			"ENTER": "\n",
		};

		if (button === "ENTER") {
			await this.sendKeys("\n");
			return;
		}

		const mappedButton = buttonMap[button];
		if (!mappedButton) {
			throw new ActionableError(`Button "${button}" is not supported on BrowserStack`);
		}

		// Use mobile:pressButton for iOS or keyevent for Android
		try {
			await this.request("POST", "/appium/device/press_button", {
				name: mappedButton,
			});
		} catch {
			// Fallback for Android
			const keyCodeMap: Record<string, number> = {
				"home": 3,
				"back": 4,
				"volumeup": 24,
				"volumedown": 25,
			};
			const keyCode = keyCodeMap[mappedButton];
			if (keyCode) {
				await this.request("POST", "/appium/device/press_keycode", {
					keycode: keyCode,
				});
			}
		}
	}

	public async tap(x: number, y: number): Promise<void> {
		await this.request("POST", "/actions", {
			actions: [
				{
					type: "pointer",
					id: "finger1",
					parameters: { pointerType: "touch" },
					actions: [
						{ type: "pointerMove", duration: 0, x, y },
						{ type: "pointerDown", button: 0 },
						{ type: "pause", duration: 100 },
						{ type: "pointerUp", button: 0 },
					],
				},
			],
		});
	}

	public async doubleTap(x: number, y: number): Promise<void> {
		await this.request("POST", "/actions", {
			actions: [
				{
					type: "pointer",
					id: "finger1",
					parameters: { pointerType: "touch" },
					actions: [
						{ type: "pointerMove", duration: 0, x, y },
						{ type: "pointerDown", button: 0 },
						{ type: "pause", duration: 50 },
						{ type: "pointerUp", button: 0 },
						{ type: "pause", duration: 100 },
						{ type: "pointerDown", button: 0 },
						{ type: "pause", duration: 50 },
						{ type: "pointerUp", button: 0 },
					],
				},
			],
		});
	}

	public async longPress(x: number, y: number, duration: number): Promise<void> {
		await this.request("POST", "/actions", {
			actions: [
				{
					type: "pointer",
					id: "finger1",
					parameters: { pointerType: "touch" },
					actions: [
						{ type: "pointerMove", duration: 0, x, y },
						{ type: "pointerDown", button: 0 },
						{ type: "pause", duration },
						{ type: "pointerUp", button: 0 },
					],
				},
			],
		});
	}

	public async getElementsOnScreen(): Promise<ScreenElement[]> {
		// Get page source and parse elements
		await this.request("GET", "/source");

		// Parse XML/JSON source to extract elements
		// BrowserStack returns Appium-format source
		const elements: ScreenElement[] = [];

		try {
			// Try to find elements using Appium's find elements
			const allElements = await this.request("POST", "/elements", {
				using: "xpath",
				value: "//*[@clickable='true' or @type='XCUIElementTypeButton' or @type='XCUIElementTypeTextField' or @type='XCUIElementTypeStaticText']",
			});

			if (allElements?.value) {
				for (const elem of allElements.value) {
					const elementId = elem.ELEMENT || elem["element-6066-11e4-a52e-4f735466cecf"];
					if (!elementId) {continue;}

					try {
						const [rect, text, label] = await Promise.all([
							this.request("GET", `/element/${elementId}/rect`),
							this.request("GET", `/element/${elementId}/text`).catch(() => ({ value: "" })),
							this.request("GET", `/element/${elementId}/attribute/label`).catch(() => ({ value: "" })),
						]);

						elements.push({
							type: "element",
							text: text?.value || "",
							label: label?.value || "",
							rect: {
								x: rect.value.x,
								y: rect.value.y,
								width: rect.value.width,
								height: rect.value.height,
							},
						});
					} catch {
						// Skip elements that can't be inspected
					}
				}
			}
		} catch (error) {
			// Fallback: return empty array if element inspection fails
		}

		return elements;
	}

	public async setOrientation(orientation: Orientation): Promise<void> {
		await this.request("POST", "/orientation", {
			orientation: orientation.toUpperCase(),
		});
	}

	public async getOrientation(): Promise<Orientation> {
		const result = await this.request("GET", "/orientation");
		return result.value.toLowerCase() as Orientation;
	}

	/**
	 * Close the BrowserStack session
	 */
	public async close(): Promise<void> {
		await this.request("DELETE", "");
	}
}

/**
 * Parse a BrowserStack device ID back to its components
 */
export function parseBrowserStackDeviceId(deviceId: string): { os: string; device: string; osVersion: string } | null {
	if (!deviceId.startsWith("browserstack:")) {
		return null;
	}

	const parts = deviceId.split(":");
	if (parts.length !== 4) {
		return null;
	}

	return {
		os: parts[1],
		device: parts[2],
		osVersion: parts[3],
	};
}

File: src/mobile_mcp/robots/browserstack.py
"""BrowserStack Robot - controls remote devices via BrowserStack's Appium hub.

Implements the Robot protocol for BrowserStack App Automate devices.
Uses WebDriver W3C Actions for touch interactions.
"""

from __future__ import annotations

from typing import Any

from ..clients.browserstack_api import (
    BrowserStackClient,
    BrowserStackCredentials,
)
from .base import (
    ActionableError,
    BaseRobot,
    InstalledApp,
    Orientation,
    Rect,
    ScreenElement,
    ScreenSize,
    SwipeDirection,
)


class BrowserStackRobot(BaseRobot):
    """Robot implementation for BrowserStack cloud devices.

    Uses BrowserStack's Appium hub to control remote real devices.
    Supports both iOS and Android platforms.
    """

    def __init__(
        self,
        credentials: BrowserStackCredentials,
        session_id: str,
        client: BrowserStackClient | None = None,
    ) -> None:
        """Initialize BrowserStack robot.

        Args:
            credentials: BrowserStack authentication credentials
            session_id: Active Appium session ID
            client: Optional BrowserStack client (created if not provided)
        """
        self.credentials = credentials
        self.session_id = session_id
        self._client = client or BrowserStackClient(credentials)
        self._owns_client = client is None

    async def _request(
        self,
        method: str,
        path: str,
        body: dict[str, Any] | None = None,
    ) -> Any:
        """Make a request to the BrowserStack session."""
        return await self._client.session_request(
            self.session_id, method, path, body
        )

    async def close(self) -> None:
        """Close the BrowserStack session and client."""
        try:
            await self._request("DELETE", "")
        finally:
            if self._owns_client:
                await self._client.close()

    async def get_screen_size(self) -> ScreenSize:
        """Get the screen dimensions."""
        result = await self._request("GET", "/window/rect")
        return ScreenSize(
            width=result["value"]["width"],
            height=result["value"]["height"],
            scale=1.0,  # BrowserStack returns actual pixels
        )

    async def get_screenshot(self) -> bytes:
        """Capture a screenshot as PNG bytes."""
        import base64

        result = await self._request("GET", "/screenshot")
        return base64.b64decode(result["value"])

    async def tap(self, x: int, y: int) -> None:
        """Tap at the specified coordinates."""
        await self._request(
            "POST",
            "/actions",
            {
                "actions": [
                    {
                        "type": "pointer",
                        "id": "finger1",
                        "parameters": {"pointerType": "touch"},
                        "actions": [
                            {"type": "pointerMove", "duration": 0, "x": x, "y": y},
                            {"type": "pointerDown", "button": 0},
                            {"type": "pause", "duration": 100},
                            {"type": "pointerUp", "button": 0},
                        ],
                    }
                ]
            },
        )

    async def double_tap(self, x: int, y: int) -> None:
        """Double-tap at the specified coordinates."""
        await self._request(
            "POST",
            "/actions",
            {
                "actions": [
                    {
                        "type": "pointer",
                        "id": "finger1",
                        "parameters": {"pointerType": "touch"},
                        "actions": [
                            {"type": "pointerMove", "duration": 0, "x": x, "y": y},
                            {"type": "pointerDown", "button": 0},
                            {"type": "pause", "duration": 50},
                            {"type": "pointerUp", "button": 0},
                            {"type": "pause", "duration": 100},
                            {"type": "pointerDown", "button": 0},
                            {"type": "pause", "duration": 50},
                            {"type": "pointerUp", "button": 0},
                        ],
                    }
                ]
            },
        )

    async def long_press(self, x: int, y: int, duration_ms: int) -> None:
        """Long press at the specified coordinates."""
        await self._request(
            "POST",
            "/actions",
            {
                "actions": [
                    {
                        "type": "pointer",
                        "id": "finger1",
                        "parameters": {"pointerType": "touch"},
                        "actions": [
                            {"type": "pointerMove", "duration": 0, "x": x, "y": y},
                            {"type": "pointerDown", "button": 0},
                            {"type": "pause", "duration": duration_ms},
                            {"type": "pointerUp", "button": 0},
                        ],
                    }
                ]
            },
        )

    async def _perform_swipe(
        self,
        start_x: int,
        start_y: int,
        end_x: int,
        end_y: int,
        duration: int = 1000,
    ) -> None:
        """Perform a swipe gesture."""
        await self._request(
            "POST",
            "/actions",
            {
                "actions": [
                    {
                        "type": "pointer",
                        "id": "finger1",
                        "parameters": {"pointerType": "touch"},
                        "actions": [
                            {
                                "type": "pointerMove",
                                "duration": 0,
                                "x": start_x,
                                "y": start_y,
                            },
                            {"type": "pointerDown", "button": 0},
                            {
                                "type": "pointerMove",
                                "duration": duration,
                                "x": end_x,
                                "y": end_y,
                            },
                            {"type": "pointerUp", "button": 0},
                        ],
                    }
                ]
            },
        )

        # Clear actions
        await self._request("DELETE", "/actions")

    async def swipe(self, direction: SwipeDirection) -> None:
        """Swipe from the center of the screen."""
        screen_size = await self.get_screen_size()
        center_x = screen_size.width // 2
        center_y = screen_size.height // 2
        distance = int(screen_size.height * 0.4)

        start_x = center_x
        start_y = center_y
        end_x = center_x
        end_y = center_y

        if direction == SwipeDirection.UP:
            start_y = center_y + distance // 2
            end_y = center_y - distance // 2
        elif direction == SwipeDirection.DOWN:
            start_y = center_y - distance // 2
            end_y = center_y + distance // 2
        elif direction == SwipeDirection.LEFT:
            start_x = center_x + distance // 2
            end_x = center_x - distance // 2
        elif direction == SwipeDirection.RIGHT:
            start_x = center_x - distance // 2
            end_x = center_x + distance // 2

        await self._perform_swipe(start_x, start_y, end_x, end_y)

    async def swipe_from_coordinate(
        self,
        x: int,
        y: int,
        direction: SwipeDirection,
        distance: int | None = None,
    ) -> None:
        """Swipe from specific coordinates."""
        swipe_distance = distance or 400

        end_x = x
        end_y = y

        if direction == SwipeDirection.UP:
            end_y = y - swipe_distance
        elif direction == SwipeDirection.DOWN:
            end_y = y + swipe_distance
        elif direction == SwipeDirection.LEFT:
            end_x = x - swipe_distance
        elif direction == SwipeDirection.RIGHT:
            end_x = x + swipe_distance

        await self._perform_swipe(x, y, end_x, end_y)

    async def send_keys(self, text: str) -> None:
        """Type text into the focused element."""
        # Try to find and send keys to active element
        try:
            active_element = await self._request("POST", "/element/active")
            if active_element and active_element.get("value", {}).get("ELEMENT"):
                element_id = active_element["value"]["ELEMENT"]
                await self._request(
                    "POST",
                    f"/element/{element_id}/value",
                    {"value": list(text)},
                )
                return
        except Exception:
            pass

        # Fallback: use keyboard actions
        key_actions = []
        for char in text:
            key_actions.append({"type": "keyDown", "value": char})
            key_actions.append({"type": "keyUp", "value": char})

        await self._request(
            "POST",
            "/actions",
            {
                "actions": [
                    {
                        "type": "key",
                        "id": "keyboard",
                        "actions": key_actions,
                    }
                ]
            },
        )

    async def press_button(self, button: str) -> None:
        """Press a hardware or software button."""
        button_map = {
            "HOME": "home",
            "BACK": "back",
            "VOLUME_UP": "volumeup",
            "VOLUME_DOWN": "volumedown",
            "ENTER": "\n",
        }

        if button == "ENTER":
            await self.send_keys("\n")
            return

        mapped_button = button_map.get(button)
        if not mapped_button:
            raise ActionableError(f'Button "{button}" is not supported on BrowserStack')

        # Try iOS-style button press first
        try:
            await self._request(
                "POST",
                "/appium/device/press_button",
                {"name": mapped_button},
            )
            return
        except Exception:
            pass

        # Fallback for Android: use key codes
        key_code_map = {
            "home": 3,
            "back": 4,
            "volumeup": 24,
            "volumedown": 25,
        }
        key_code = key_code_map.get(mapped_button)
        if key_code:
            await self._request(
                "POST",
                "/appium/device/press_keycode",
                {"keycode": key_code},
            )

    async def list_apps(self) -> list[InstalledApp]:
        """List installed applications.

        Note: BrowserStack doesn't provide an API to list installed apps.
        Returns an empty list - user should know which apps they've uploaded.
        """
        return []

    async def launch_app(self, package_name: str, locale: str | None = None) -> None:
        """Launch an application by package/bundle ID."""
        await self._request(
            "POST",
            "/appium/device/activate_app",
            {
                "bundleId": package_name,  # iOS
                "appPackage": package_name,  # Android - Appium handles both
            },
        )

    async def terminate_app(self, package_name: str) -> None:
        """Terminate a running application."""
        await self._request(
            "POST",
            "/appium/device/terminate_app",
            {
                "bundleId": package_name,
                "appPackage": package_name,
            },
        )

    async def install_app(self, path: str) -> None:
        """Install an application.

        Note: For BrowserStack, apps must be uploaded first via the API.
        The path should be a bs:// URL from a previous upload.
        """
        if path.startswith("bs://"):
            await self._request(
                "POST",
                "/appium/device/install_app",
                {"appPath": path},
            )
        else:
            raise ActionableError(
                "BrowserStack requires apps to be uploaded first. "
                "Use the BrowserStack API to upload your app and get a bs:// URL."
            )

    async def uninstall_app(self, bundle_id: str) -> None:
        """Uninstall an application."""
        await self._request(
            "POST",
            "/appium/device/remove_app",
            {
                "bundleId": bundle_id,
                "appPackage": bundle_id,
            },
        )

    async def open_url(self, url: str) -> None:
        """Open a URL or deep link."""
        await self._request("POST", "/url", {"url": url})

    async def get_elements_on_screen(self) -> list[ScreenElement]:
        """Get all UI elements currently visible on screen."""
        elements: list[ScreenElement] = []

        try:
            # Find interactive elements using XPath
            all_elements = await self._request(
                "POST",
                "/elements",
                {
                    "using": "xpath",
                    "value": (
                        "//*[@clickable='true' or "
                        "@type='XCUIElementTypeButton' or "
                        "@type='XCUIElementTypeTextField' or "
                        "@type='XCUIElementTypeStaticText']"
                    ),
                },
            )

            if all_elements and all_elements.get("value"):
                for elem in all_elements["value"]:
                    element_id = elem.get("ELEMENT") or elem.get(
                        "element-6066-11e4-a52e-4f735466cecf"
                    )
                    if not element_id:
                        continue

                    try:
                        # Get element properties
                        rect = await self._request(
                            "GET", f"/element/{element_id}/rect"
                        )

                        # Try to get text and label
                        try:
                            text_result = await self._request(
                                "GET", f"/element/{element_id}/text"
                            )
                            text = text_result.get("value", "")
                        except Exception:
                            text = ""

                        try:
                            label_result = await self._request(
                                "GET", f"/element/{element_id}/attribute/label"
                            )
                            label = label_result.get("value", "")
                        except Exception:
                            label = ""

                        elements.append(
                            ScreenElement(
                                type="element",
                                text=text or None,
                                label=label or None,
                                rect=Rect(
                                    x=rect["value"]["x"],
                                    y=rect["value"]["y"],
                                    width=rect["value"]["width"],
                                    height=rect["value"]["height"],
                                ),
                            )
                        )
                    except Exception:
                        # Skip elements that can't be inspected
                        continue

        except Exception:
            # Return empty array if element inspection fails
            pass

        return elements

    async def get_orientation(self) -> Orientation:
        """Get the current screen orientation."""
        result = await self._request("GET", "/orientation")
        orientation_str = result["value"].lower()
        return Orientation(orientation_str)

    async def set_orientation(self, orientation: Orientation) -> None:
        """Set the screen orientation."""
        await self._request(
            "POST",
            "/orientation",
            {"orientation": orientation.value.upper()},
        )
